python/ray/experimental/async_plasma.py

```diff
@@ -0,0 +1,235 @@
+import asyncio
+import sys
+from typing import List, Dict
+
+import pyarrow.plasma as plasma
+
+import ray
+from ray.services import logger
+
+
+def _release_waiter(waiter, *_):
+    if not waiter.done():
+        waiter.set_result(None)
+
+
+class PlasmaProtocol(asyncio.Protocol):
+    """
+    Protocol control for the asyncio connection.
+    """
+    def __init__(self, plasma_client,
+                 plasma_event_handler: 'PlasmaEventHandler'):
+        self.plasma_client = plasma_client
+        self.plasma_event_handler = plasma_event_handler
+        self.transport = None
+        self._buffer = b''
+
+    def connection_made(self, transport):
+        self.transport = transport
+
+    def data_received(self, data):
+        self._buffer += data
+        messages = []
+        i = 0
+        while i + 8 <= len(self._buffer):
+            msg_len = int.from_bytes(self._buffer[i:i + 8], sys.byteorder)
+            if i + 8 + msg_len > len(self._buffer):
+                break
+            i += 8
+            segment = self._buffer[i: i + msg_len]
+            i += msg_len
+            messages.append(self.plasma_client.decode_notification(segment))
+
+        self._buffer = self._buffer[i:]
+        self.plasma_event_handler.process_notifications(messages)
+
+    def connection_lost(self, exc):
+        # The socket has been closed
+        logger.debug('PlasmaProtocol - connection lost.')
+
+    def eof_received(self):
+        logger.debug('PlasmaProtocol - EOF received.')
+        self.transport.close()
+
+
+class PlasmaObjectFuture(asyncio.Future):
+    """This class manages the lifecycle of a Future contains an object_id.
+
+    Note:
+        This Future is an item in an linked list.
+
+    Attributes:
+        object_id: The object_id this Future contains.
+    """
+
+    def __init__(self, loop, object_id):
+        super().__init__(loop=loop)
+        self.object_id = object_id
+        self.prev = None
+        self.next = None
+
+    @property
+    def ray_object_id(self):
+        return ray.ObjectID(self.object_id.binary())
+
+    def __repr__(self):
+        return super().__repr__() + "{object_id=%s}" % self.object_id
+
+
+class PlasmaObjectLinkedList(asyncio.Future):
```

Comment:
Do we need this now that we just have as_future()?

Reply:
This is a data structure for event handler, not the futures for ObjectIDs.